### ISSUE TITLE
mkinitcpio hook / GRUB not playing along

### ISSUE THREAD

**Issue by UweSauter** [archzfs/archzfs#228]
I don't know whether this is an issue of incomplete documentation in the [Installing Arch Linux on ZFS](https://wiki.archlinux.org/index.php/Installing_Arch_Linux_on_ZFS) article in the Arch wiki, an issue with the current version of GRUB (2.02-4) or with the mkinitcpio zfs hook script (which is provided by zfs-utils-common-git). So feel free to close this if you think this should be handled in the documentation or by GRUB.

I freshly installed a system following the Arch wiki article. These are pool properties and layout:

```
# zpool get all | sort 
NAME   PROPERTY                       VALUE                          SOURCE
zroot  allocated                      997M                           -
zroot  altroot                        -                              default
zroot  ashift                         12                             local
zroot  autoexpand                     off                            default
zroot  autoreplace                    off                            default
zroot  bootfs                         zroot/SYSTEM/default           local
zroot  cachefile                      -                              default
zroot  capacity                       0%                             -
zroot  comment                        -                              default
zroot  dedupditto                     0                              default
zroot  dedupratio                     1.00x                          -
zroot  delegation                     on                             default
zroot  expandsize                     -                              -
zroot  failmode                       continue                       local
zroot  feature@async_destroy          enabled                        local
zroot  feature@bookmarks              enabled                        local
zroot  feature@edonr                  disabled                       local
zroot  feature@embedded_data          active                         local
zroot  feature@empty_bpobj            active                         local
zroot  feature@enabled_txg            active                         local
zroot  feature@encryption             enabled                        local
zroot  feature@extensible_dataset     active                         local
zroot  feature@filesystem_limits      enabled                        local
zroot  feature@hole_birth             active                         local
zroot  feature@large_blocks           enabled                        local
zroot  feature@large_dnode            disabled                       local
zroot  feature@lz4_compress           active                         local
zroot  feature@multi_vdev_crash_dump  disabled                       local
zroot  feature@sha512                 disabled                       local
zroot  feature@skein                  disabled                       local
zroot  feature@spacemap_histogram     active                         local
zroot  feature@userobj_accounting     active                         local
zroot  fragmentation                  0%                             -
zroot  free                           463G                           -
zroot  freeing                        0                              -
zroot  guid                           2084561213606389230            -
zroot  health                         ONLINE                         -
zroot  leaked                         0                              -
zroot  listsnapshots                  off                            default
zroot  multihost                      off                            default
zroot  readonly                       off                            -
zroot  size                           464G                           -
zroot  version                        -                              default

# zfs list
NAME                   USED  AVAIL  REFER  MOUNTPOINT
zroot                 5.22G   444G    96K  none
zroot/DATA             192K   444G    96K  none
zroot/DATA/home         96K   444G    96K  /home
zroot/SYSTEM           993M   444G    96K  none
zroot/SYSTEM/default   993M   444G   993M  /
zroot/swap            4.25G   449G    56K  -
```

The first reboot (and many more…) got stuck with:

```
  Booting `Arch Linux'
                      
Loading Linux linux ...
Loading initial ramdisk ...
[    0.394383] ERST: Failed to get Error Log Address Range.
[    0.494905] APEI: Can not request [mem 0xbff9a740-0xbff9a793] for APEI BERT registers
:: running early hook [udev]
starting version 237
:: running hook [udev]
:: Triggering uevents...
:: running hook [keymap]
:: Loading keymap...kbd_mode: KDSKBMODE: Inappropriate ioctl for device
Gdone.
:: running hook [consolefont]
:: Loading console font...setfont: KDFONTOP: Inappropriate ioctl for device
done.
:: running hook [zfs]
:: mounting 'ZFS=zroot/SYSTEM/default' on real root
filesystem 'ZFS=zroot/SYSTEM/default' cannot be mounted, unable to open the dataset
You are now being dropped into an emergency shell.
sh: can't access tty; job control turned off
```

I was able to get the system to boot completely by:

```
[rootfs ]# zpool import -R /new_root zroot

[rootfs ]# exit

Trying to continue (this will most likely fail) ...
:: running late hook [zfs]
no pools available to import
:: running cleanup hook [udev]

Welcome to Arch Linux!

[  OK  ] Created slice System Slice.
```

As far as I can tell the problem is that GRUB creates the following command line (taken from /proc/cmdline of the successfully booted system)

`BOOT_IMAGE=/SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw`

and that the mkinitcpio hook script doesn't know how to handle `root=ZFS=…` As far as I got reading the hook script it only looks for `zfs=…` and thus doesn't import / mount the pool.

If I edit `/etc/default/grub` to include `zfs=bootfs` in `GRUB_CMDLINE_LINUX_DEFAULT` and recreate `/boot/grub/grub.cfg` then the boot process runs successfully but the kernel command line then is kind of redundant `BOOT_IMAGE=/SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw zfs=bootfs`:

```
  Booting `Arch Linux'
                      
Loading Linux linux ...
Loading initial ramdisk ...
[    0.395755] ERST: Failed to get Error Log Address Range.
[    0.495959] APEI: Can not request [mem 0xbff9a740-0xbff9a793] for APEI BERT registers
:: running early hook [udev]
starting version 237
:: running hook [udev]
:: Triggering uevents...
:: running hook [keymap]
:: Loading keymap...kbd_mode: KDSKBMODE: Inappropriate ioctl for device
Gdone.
:: running hook [consolefont]
:: Loading console font...setfont: KDFONTOP: Inappropriate ioctl for device
done.
:: running hook [zfs]
:: running late hook [zfs]
no pools available to import
:: running cleanup hook [udev]

Welcome to Arch Linux!
```

Another workaround is to edit `/boot/grub/grub.cfg` and remove `root=` so that there is only `ZFS=…`

If the hook could also react on `root=ZFS=…` this would avoid the need to edit `/boot/grub/grub.cfg` after it is recreated (which might happen from time to time) *or* to have the redundant `zfs=bootfs` or `zfs=<poolname>` in `/etc/default/grub`.

What do you think?

**Comment by minextu** [comment]
Thanks for reporting this.
https://github.com/minextu/archzfs/commit/cba22f09671c5f49796f2b58d681354561995e98 should make it work, but I haven't had time to test yet.

**Comment by UweSauter** [comment]
I'll test it manually tomorrow.

**Comment by UweSauter** [comment]
Testing your proposed changes yield:

*working* for
**/proc/cmdline:**
`BOOT_IMAGE=/SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw zfs=bootfs`
**grub.cfg** (with "zfs=bootfs" in `/etc/default/grub`):
`linux   /SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw zfs=bootfs`

*working* for
**/proc/cmdline:**
`BOOT_IMAGE=/SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw`
**grub.cfg** (without "zfs=bootfs" in `/etc/default/grub`):
`linux   /SYSTEM/default@/boot/vmlinuz-linux root=ZFS=zroot/SYSTEM/default rw`

*working* for
**/proc/cmdline:**
`BOOT_IMAGE=/SYSTEM/default@/boot/vmlinuz-linux zfs=bootfs rw`
**grub.cfg** (manually removed `root=…`):
`linux   /SYSTEM/default@/boot/vmlinuz-linux zfs=bootfs rw`

**Comment by killhellokitty** [comment]
This is not fixed, at least for the -git versions.
First grub.cfg  cuts off the pool name on the root=ZFS=/SYSTEM/default rw for example.
Secondly I have had to change root=ZFS=. In the grub.cfg to zfs='add zpool name here'/SYSTEM/default rw and add root='zpool name'/SYSTEM/default to the grub kernel cnd line or the system will not boot.
I had spent many hours trying to find a combination that woul boot.
This is currently an issue at least when using the zfs-git pkgs
